libcxx/include/__type_traits/diagnostic_utilities.h

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___TYPE_TRAITS_DIAGNOSTIC_UTILITIES_H
+#define _LIBCPP___TYPE_TRAITS_DIAGNOSTIC_UTILITIES_H
+
+#include <__config>
+#include <__type_traits/decay.h>
+#include <__type_traits/integral_constant.h>
+#include <__type_traits/is_bounded_array.h>
+#include <__type_traits/is_const.h>
+#include <__type_traits/is_function.h>
+#include <__type_traits/is_reference.h>
+#include <__type_traits/is_same.h>
+#include <__type_traits/is_unbounded_array.h>
+#include <__type_traits/is_void.h>
+#include <__type_traits/is_volatile.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+// Many templates require their type parameters to be cv-unqualified objects.
+template <template <class...> class _Template, class _Tp, bool = is_same<__decay_t<_Tp>, _Tp>::value>
+struct __requires_cv_unqualified_object_type : true_type {};
+
+#define _LIBCPP_DEFINE__REQUIRES_CV_UNQUALIFIED_OBJECT_TYPE(_Template, _Verb)                                          \
+  template <class _Tp>                                                                                                 \
+  struct __requires_cv_unqualified_object_type<_Template, _Tp, false>                                                  \
```

Comment:
Is there a reason why the patch now defines these `__requires_cv_unqualified_object_type` specializations for each container? Was that in response to the comment about too many instantiations?

This adds complexity but I don't see the benefit (yet), I'd rather keep this patch as simple as can be. After all the goal here is to issue diagnostics -- that's a fairly simple problem and I'd like to keep the solution accordingly simple.

Reply:
> Is there a reason why the patch now defines these `__requires_cv_unqualified_object_type` specializations for each container? Was that in response to the comment about too many instantiations?

Yes. I ended up benchmarking Chrome's build times and while we could probably tolerate this patch in isolation, stacking similar changes will eventually tank build times. 

> This adds complexity but I don't see the benefit (yet), I'd rather keep this patch as simple as can be. After all the goal here is to issue diagnostics -- that's a fairly simple problem and I'd like to keep the solution accordingly simple.

Issuing simple diagnostics is simple. The goal of this patch is to issue diagnostics that are informative to the reader, using language that they're likely to understand; achieving that without adversely impacting build times is apparently more challenging to simplify.

Could you help me understand what it is about the complexity that raises concern, please?

Reply:
Basically, I don't understand why the macros don't expand to a bunch of `static_asserts` directly (like it used to do), instead of going through an additional partial specialization of `__requires_cv_unqualified_object_type` & friends. That's a complex mechanism for doing basically `static_assert(condition)` and I don't understand what we're gaining from that complexity. In fact, I would expect this approach to have more impact on build times since we instantiate more stuff *and* we have a partial specialization.

Reply:
Ah, I get where you're coming from now, thank you. This is a metaprogramming hack to preserve performance for valid use.
* In the case where `decay_t<_Tp>` matches `_Tp`: `__requires_cv_unqualified_type` should only need to instantiate two templates: `decay_t<_Tp>` and `__requires_cv_unqualified_type<_Template, _Tp, true>`.
* In case where `decay_t<_Tp>` is different to `_Tp`: `__requires_cv_unqualified_type` then instantiates all the checks to see why that isn't the case. This is indeed slower, but we're now going for a failed build, and tooling tends to burn compile times in favour of helpful diagnostics.

Neither `__allocator_requirements`, nor `__container_requirements` have the same effect, so I'm happy to revert those without further convincing.

Chromium is a good candidate for benchmarking the worst-case scenario, because it has well over 100k utterances of `std::vector`. Would providing benchmarks be helpful? (It'll take ~a day to get those as I discarded the results from last week, and I tend to run quite a few trials to reduce the impact of noise.)

Reply:
Ping

Reply:
Ping :)

Reply:
I'm skeptical about the performance claims since the current solution requires instantiating `is_same`, `__decay_t`, `__requires_cv_unqualified_object_type` and finding the right specialization for it. That doesn't seem to be significantly less expensive than instantiating the 4 traits directly. And if we wanted to be even speedier, we could use the `_v` versions of these traits (which would bump requirements to C++17), or we could use the compiler builtins directly for `__is_reference` and `__is_function`, which we use unconditionally.

And even if there's a very very small compile-time hit, the complexity introduced by the current approach for what is essentially 4 `static_assert`s seems really difficult to justify.

I would have no opposition to this patch if it was simplified to something like:

```c++
#if _LIBCPP_STD_VER >= 20
# define _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type)                                                   \
    static_assert(!is_const<value_type>::value, #template " cannot have a const value_type");                         \
    static_assert(!is_volatile<value_type>::value, #template " cannot have a volatile value_type");                   \
    static_assert(!is_reference<value_type>::value, #template " cannot have a value_type that is a reference type");  \
    static_assert(!is_function<value_type>::value, #template " cannot have a value_type that is a function type")
#else
# define _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type)                                                   \
    static_assert(!is_const<value_type>::value, #template " cannot have a const value_type");                         \
    static_assert(!is_volatile<value_type>::value, #template " cannot have a volatile value_type");                   \
    static_assert(!is_reference<value_type>::value, #template " cannot have a value_type that is a reference type");  \
    static_assert(!is_function<value_type>::value, #template " cannot have a value_type that is a function type");    \
    static_assert(!__libcpp_is_bounded_array<value_type>::value, #template " cannot have a value_type that is a bounded array before C++20")
#endif

#define _LIBCPP_CHECK_ALLOCATOR_AWARE_CONTAINER_REQUIREMENTS(template, value_type)                                    \
  _LIBCPP_CHECK_ALLOCATOR_REQUIREMENTS(template, value_type);                                                         \
  static_assert(!is_void<value_type>::value, #template " cannot hold void");                                          \
  static_assert(!__libcpp_is_unbounded_array<value_type>::value, #template " cannot hold C arrays of an unknown size")
```

I think it's difficult to justify adding more complexity than that either for compilation times or for diagnostics quality. After all, this kind of error is something that users will see once when they try to start using a container wrong, and they will very likely see this near the top of their call stack (i.e. this is unlikely to trigger deep down inside a dependent library). 95% of the benefit of this patch is obtained by adding the `static_assert`s in the simplest way possible, but as it stands a large part of the complexity comes from trying to go the last 5% in terms of diagnostics quality and compilation time (which I'm still skeptical of). I don't think that's the right tradeoff, and that's the source of my pushback.

So, TLDR: I think adding these diagnostics is great and I love that we're making the library more robust to misuse and improving the diagnostics quality. But I don't think the patch achieves the right complexity tradeoff in its current state (it started more naive and with a better tradeoff IMO).